opnode/l1/source.go

```diff
@@ -4,161 +4,335 @@ import (
-		Parent: eth.BlockID{Hash: header.ParentHash, Number: parentNum},
-	}, nil
-}
-
-// L1Range returns a range of L1 block beginning just after `begin`.
-func (s Source) L1Range(ctx context.Context, begin eth.BlockID) ([]eth.BlockID, error) {
-	// Ensure that we start on the expected chain.
-	if canonicalBegin, err := s.L1BlockRefByNumber(ctx, begin.Number); err != nil {
-		return nil, fmt.Errorf("failed to fetch L1 block %v %v: %w", begin.Number, begin.Hash, err)
-	} else {
-		if canonicalBegin.Self != begin {
-			return nil, fmt.Errorf("Re-org at begin block. Expected: %v. Actual: %v", begin, canonicalBegin.Self)
+	for i := 0; i < len(blockRequests); i++ {
+		if blockRequests[i].Error != nil {
+			return nil, fmt.Errorf("failed to retrieve transactions of block %s in batch of %d blocks: %v", window[i], len(blockRequests), blockRequests[i].Error)
 		}
 	}
 
-	l1head, err := s.L1HeadBlockRef(ctx)
+	return allTxLists, nil
+}
+
+func (s *Source) refFromStart(ctx context.Context, info derive.L1Info) (eth.L1BlockRef, error) {
+	if info.NumberU64() == 0 {
+		return eth.L1BlockRef{Self: info.ID()}, nil // L1 genesis block has zeroed parent hash / number
+	}
+	parent, err := s.InfoByHash(ctx, info.ParentHash())
 	if err != nil {
-		return nil, fmt.Errorf("failed to fetch head L1 block: %w", err)
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch parent of %s: %v", info.ID(), err)
 	}
-	maxBlocks := MaxBlocksInL1Range
-	// Cap maxBlocks if there are less than maxBlocks between `begin` and the head of the chain.
-	if l1head.Self.Number-begin.Number <= maxBlocks {
-		maxBlocks = l1head.Self.Number - begin.Number
+	return eth.L1BlockRef{Self: info.ID(), Parent: parent.ID()}, nil
+}
+
+func (s *Source) L1HeadBlockRef(ctx context.Context) (eth.L1BlockRef, error) {
+	head, err := s.InfoHead(ctx)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch head header: %v", err)
 	}
+	return s.refFromStart(ctx, head)
+}
 
-	if maxBlocks == 0 {
-		return nil, nil
+func (s *Source) L1BlockRefByNumber(ctx context.Context, l1Num uint64) (eth.L1BlockRef, error) {
+	head, err := s.InfoByNumber(ctx, l1Num)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch header by num %d: %v", l1Num, err)
 	}
+	return s.refFromStart(ctx, head)
+}
 
-	prevHash := begin.Hash
-	var res []eth.BlockID
-	// TODO: Walk backwards to be able to use block by hash
-	for i := begin.Number + 1; i < begin.Number+maxBlocks+1; i++ {
-		n, err := s.L1BlockRefByNumber(ctx, i)
-		if err != nil {
-			return nil, fmt.Errorf("failed to fetch L1 block %v: %w", i, err)
+// L1Range returns a range of L1 block beginning just after begin, up to max blocks.
+// This batch-requests all blocks by number in the range at once, and then verifies the consistency
+func (s *Source) L1Range(ctx context.Context, begin eth.BlockID, max uint64) ([]eth.BlockID, error) {
+
+	headerRequests := make([]rpc.BatchElem, max)
+	for i := uint64(0); i < max; i++ {
```

Comment:
Currently it fetches everything by number, and then returns the block ids after checking the headers form one canonical chain. If it reorgs or misses anything at any point in the batch response, it will just return the prefix of the range (function may return any number of blocks to add to the existing buffer). I think this is fastest, because the batch can be optimized and fully parallel if we configure it that way.